src/storage/hummock_trace/src/record.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::ops::Bound;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use bincode::error::{DecodeError, EncodeError};
+use bincode::{Decode, Encode};
+use prost::Message;
+use risingwave_pb::meta::SubscribeResponse;
+
+use crate::StorageType;
+
+pub type RecordId = u64;
+
+pub(crate) struct RecordIdGenerator {
+    record_id: AtomicU64,
+}
+
+impl RecordIdGenerator {
+    pub(crate) fn new() -> Self {
+        Self {
+            record_id: AtomicU64::new(0),
+        }
+    }
+
+    pub(crate) fn next(&self) -> RecordId {
+        self.record_id.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+#[derive(Encode, Decode, Debug, PartialEq, Clone)]
+pub struct Record(pub StorageType, pub RecordId, pub Operation);
+
+impl Record {
+    pub(crate) fn new(storage_type: StorageType, record_id: RecordId, op: Operation) -> Self {
+        Self(storage_type, record_id, op)
+    }
+
+    pub fn storage_type(&self) -> &StorageType {
+        &self.0
+    }
+
+    pub fn record_id(&self) -> RecordId {
+        self.1
+    }
+
+    pub fn op(&self) -> &Operation {
+        &self.2
+    }
+
+    pub fn is_iter_related(&self) -> bool {
+        matches!(self.op(), Operation::Iter { .. } | Operation::IterNext(_))
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_local_none(record_id: RecordId, op: Operation) -> Self {
+        Self::new(StorageType::Global, record_id, op)
+    }
+}
+
+type TraceKey = Vec<u8>;
```

Comment:
Can we use bytes::Bytes here? It is a wrapper over a piece of immutable memory, and do reference counting when cloned. In this way, we don't have to allocate and store extra memory in this trace framework.

Reply:
> I think we can always write to a separated file. We can implement an extra tools to interpret the content of the trace file. Writing to stdout can easily mix with other logs.

What about a `serde_json` implementation for `Serializer`? Make it a configurable option so it supports multiple serialization data formats.